platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/internal/component/ResolutionFailureHandlerIntegrationTest.groovy

```diff
@@ -16,42 +16,461 @@
+        noMatchingArtifactVariants.prepare()
+
+        expect:
+        fails "forceResolution", "--stacktrace"
+        failure.assertHasErrorOutput("Caused by: " + NoMatchingArtifactVariantsException.class.getName())
+        failure.assertHasDescription("Could not determine the dependencies of task ':forceResolution'.")
+        failure.assertHasCause("Could not resolve all task dependencies for configuration ':resolveMe'.")
+        assertFullMessageCorrect("No variants of project : match the consumer attributes:")
+    }
+
+    def "demonstrate ambiguous artifact transforms exception"() {
+        ambiguousArtifactTransforms.prepare()
+
+        expect:
+        fails "forceResolution", "--stacktrace"
+        failure.assertHasErrorOutput("Caused by: " + AmbiguousArtifactTransformException.class.getName())
+        failure.assertHasDescription("Could not determine the dependencies of task ':forceResolution'.")
+        failure.assertHasCause("Could not resolve all task dependencies for configuration ':resolveMe'.")
+        assertFullMessageCorrect("Found multiple transforms that can produce a variant of project : with requested attributes:")
+    }
+
+    def "demonstrate ambiguous artifact variants exception"() {
+        ambiguousArtifactVariants.prepare()
+
+        expect:
+        fails "forceResolution", "--stacktrace"
+        failure.assertHasErrorOutput("Caused by: " + AmbiguousArtifactVariantsException.class.getName())
+        failure.assertHasDescription("Could not determine the dependencies of task ':forceResolution'.")
+        failure.assertHasCause("Could not resolve all task dependencies for configuration ':resolveMe'.")
+        assertFullMessageCorrect("More than one variant of project : matches the consumer attributes:")
+    }
+    // endregion Artifact Variant failures
+    // endregion resolution failures
+
+    // region dependencyInsight failures
+    /**
+     * Running the dependencyInsight report can also generate a variant selection failure, but this
+     * does <strong>NOT</strong> cause the task to fail.
+     */
+    def "demonstrate dependencyInsight report no matching capabilities failure"() {
+        setupDependencyInsightFailure()
+
+        expect:
+        succeeds "dependencyInsight", "--configuration", "compileClasspath", "--dependency", "gson"
+
+        String basicOutput = """   Failures:
+      - Could not resolve com.google.code.gson:gson:2.8.5."""
+        String fullOutput = "          - Unable to find a variant of com.google.code.gson:gson:2.8.5 providing the requested capability com.google.code.gson:gson-test-fixtures:"
+
+        outputContains(basicOutput)
+        outputContains(fullOutput)
+    }
+    // endregion dependencyInsight failures
+
+    // region error showcase
+    //@spock.lang.Ignore("This test is used to generate a summary of all possible errors, it shouldn't usually be run as part of testing")
+    def "generate resolution failure showcase report"() {
```

Comment:
IMO generating this showcase report makes these tests more complicated than they need to be. I think this test class would be improved if we allowed Gradle to run all tests as normal (running the entire class), moving the build scripts into each individual test, and updating the asserts in each test above to just check for a single block of text. 

The advantage of this is that we get to see the change in output for changes to the error messages verbatim in the git diff

Reply:
I want to come back to this, but want to get the minimal change with the links finished first.

Reply:
🤔 I think this is a bit overwrought.  Can we remove/disable the test that runs everything before this is merged?

I generally agree with Justin that the most important thing to see in the test is what the user sees, so a test that looks like this:

```
    def "demonstrate ambiguous graph variant selection failure for project"() {
        setupAmbiguousGraphVariantFailureForProject()

        when:
        failsWithResolutionError()

        then:
        failure.assertHasDescription("Could not determine the dependencies of task ':forceResolution'.")
        failure.assertHasCause("Could not resolve all task dependencies for configuration ':resolveMe'.")
        failure.assertHasCause("Could not resolve project :.")
        failure.assertHasCause("""The consumer was configured to find attribute 'color' with value 'blue'. However we cannot choose between the following variants of project ::
  - blueRoundElements
  - blueSquareElements
All of them match the consumer attributes:
  - Variant 'blueRoundElements' capability :d1t0c:unspecified declares attribute 'color' with value 'blue':
      - Unmatched attribute:
          - Provides shape 'round' but the consumer didn't ask for it
  - Variant 'blueSquareElements' capability :d1t0c:unspecified declares attribute 'color' with value 'blue':
      - Unmatched attribute:
          - Provides shape 'square' but the consumer didn't ask for it""")

        and: "Helpful resolutions are provided"
        assertSuggestsReviewingAlgorithm()
        assertSuggestsViewingDocs("Ambiguity errors are explained in more detail at https://docs.gradle.org/${GradleVersion.current().version}/userguide/variant_model.html#sub:variant-ambiguity.")
    }
```

Is better because it clearly shows the description, causes and the (ugly) message you get now. It's OK if these tests are annoying to update because the intention is to make this message better.

Reply:
I will uncomment the `@Ignore` before merging, the report is only meant to be run manually.

Will add in the full WoT causes like this.  My thinking was I just wanted to have a "minimal identifier" of the WoT here so I could quickly iterate on the message and formatting, but we're not doing that yet.